key-rotator/manifest/manifest.go

```diff
@@ -30,6 +37,68 @@ type DataShareProcessorSpecificManifest struct {
 	PacketEncryptionKeyCSRs PacketEncryptionKeyCSRs `json:"packet-encryption-keys"`
 }
 
+// UpdateKeysConfig configures an UpdateKeys operation.
+type UpdateKeysConfig struct {
+	BatchSigningKey         key.Key // the key used for batch signing operations
+	BatchSigningKeyIDPrefix string  // the key ID prefix to use for batch signing keys
+
+	PacketEncryptionKey         key.Key // the key used for packet encryption operations
+	PacketEncryptionKeyIDPrefix string  // the key ID prefix to use for packet encryption keys
+	PacketEncryptionKeyCSRFQDN  string  // the FQDN to specify for packet encryption key CSRs
+}
+
+func (m DataShareProcessorSpecificManifest) UpdateKeys(cfg UpdateKeysConfig) (DataShareProcessorSpecificManifest, error) {
+	// Copy the current manifest, clearing any existing batch signing/packet encryption keys.
+	newM := m
+	newM.BatchSigningPublicKeys, newM.PacketEncryptionKeyCSRs = BatchSigningPublicKeys{}, PacketEncryptionKeyCSRs{}
+
+	// Update batch signing key.
+	for _, v := range cfg.BatchSigningKey {
+		kid := cfg.BatchSigningKeyIDPrefix
+		if !v.CreationTime.IsZero() {
+			kid = fmt.Sprintf("%s-%d", cfg.BatchSigningKeyIDPrefix, v.CreationTime.Unix())
+		}
+
+		var newBSPK BatchSigningPublicKey
+		if bspk, ok := m.BatchSigningPublicKeys[kid]; ok {
+			newBSPK = bspk
+		} else {
+			pkix, err := v.RawKey.PublicAsPKIX()
+			if err != nil {
+				return DataShareProcessorSpecificManifest{}, fmt.Errorf("couldn't create PKIX-encoding for batch signing key version created at %v (%d): %w", v.CreationTime.Format(time.RFC3339), v.CreationTime.Unix(), err)
+			}
+			newBSPK = BatchSigningPublicKey{PublicKey: pkix}
+		}
+		newM.BatchSigningPublicKeys[kid] = newBSPK
+	}
+
+	// Update packet encryption key.
+	for _, v := range cfg.PacketEncryptionKey {
+		if !v.Primary {
+			continue
+		}
```

Comment:
Should we error out if more than one key version is marked primary?

Reply:
Yes, but see my response to the other comment regarding validations.